### ISSUE TITLE
Bug with useCollection and where by timeStamp.

### ISSUE THREAD

**Issue by izakfilmalter** [nandorojo/swr-firestore#51]
Love using this lib so far. I am trying to make a paginated list of messages for a messaging app, and I think I stumbled on a bug. Because of the JSON serialization / maybe something else in useCollection, `where: ['time', '<', new Data(Date.now())]` doesn't return anything, but a regular get on a collection ref does.

```js
const collectionPath = `path/to/collection`
const orderBy = ['time', 'desc']

const currentDate = new Date(Date.now())

fuego.db
  .collection(collectionPath)
  .orderBy('time', 'desc')
  .where('time', '<', currentDate)
  .get()
  .then((data) => console.log(data.empty)) // data.empty === false

const { data } = useCollection(
  collectionPath,
  {
    orderBy: ['time', 'desc'],
    where: ['time', '<', currentDate],
    listen: true,
  },
)

// data always is empty array.

```

**Comment by nandorojo** [comment]
Yeah...this is a known issue. At the moment, only serializable fields are valid, since this is how we store items in the cache. It seems like #46 might be a solution to this, where you could convert your date string back into a date object in the `toFirestore` part. It’s not merged yet, but it’s in review.

**Comment by nandorojo** [comment]
On second thought, looks like that wouldn’t exactly solve this, it’s only for data manipulation. I’d have to give some thought into this. 

Does firestore allow any date string comparison for queries?

**Comment by jerber** [comment]
Yeah either you can convert the string Date back into a Date object or you can store the Dates as strings in firestore and query based on string Dates. You could also use Unix timestamps.

**Comment by izakfilmalter** [comment]
I tried a bunch of different conversations to get it working and couldn't get it. You might have better luck than I. Maybe we can change the api to something like this:
```js
where: ['time', '<', {type: 'date', value: currentDate.unix()}]
```
or
```js
where: ['time', '<', currentDate, {type: 'date'}]
```

**Comment by nandorojo** [comment]
Could you store your dates as Unix time stamps and then use that to query?

**Comment by izakfilmalter** [comment]
I could, but I have a lot of data with firebase.Timestamp in it. I didn't want to have to deal with drift between clients.

**Comment by nandorojo** [comment]
Got it. Something like this, like you proposed, seems like it would make sense:

```js
where: ['time', '<', {type: 'date', value: currentDate.unix()}]
```

It would still check if it’s just a value, but if there’s an object, it would parse it accordingly. I’m not sure if there would be any odd edge cases to an api change like that.

**Comment by izakfilmalter** [comment]
I think this would probably be safer, you don't have to refactor code that checks `where[2]`, and just add checks for `where[3]`.

```js
where: ['time', '<', currentDate, {type: 'date'}]
```

**Comment by nandorojo** [comment]
I think that makes sense to me. Would you like to contribute this in `src/use-collection`?

**Comment by AlexanderArvidsson** [comment]
@nandorojo Why was this closed?
I am unable to query for dates at all.

The following code does not work, it simply returns an empty array always.
```
  const { data: book } = useCollection<CM.Portfolio>(
    portfolio ? `portfolios/${portfolio.id}/books` : null,
    {
      listen: true,
      where: ['date', '>', startOfYear(new Date())]
    },
    {}
  );
```

However, querying directly from the firestore instance works:
```
  const ctx = useFuegoContext();
  if (portfolio) {
    ctx.fuego.db
      .collection(`portfolios/${portfolio.id}/books`)
      .where('date', '>', startOfYear(new Date()))
      .get()
      .then((data) => console.log(data.docs.map((d) => d.data())));
  }
```

My dates are stored as timestamps in the database.
None of the suggested syntax you mentioned works.

**Comment by nandorojo** [comment]
I can re-open if someone wants to submit a PR with [this solution](https://github.com/nandorojo/swr-firestore/issues/51#issuecomment-711093577). Otherwise, I'll try to get to it when I have time.

**Comment by AlexanderArvidsson** [comment]
I can try implementing it, but I would like to know the reason we cannot just query with the Date object, like firestore does? Is it because of swr cache like you mentioned earlier? Could we not use some kind of serializer / deserializer for date objects when storing in cache instead?

It seems odd that the API for this package would differ from the firestore instance API.

**Comment by nandorojo** [comment]
It's because this library has to stringify the query before passing to SWR in order to have a stable key.

**Comment by AlexanderArvidsson** [comment]
> Maybe we could check if it's a date object before serializing and inject a `type: date` under the hood if it is.

I personally think this is the better solution. If the user specifically provides a `where[3]`, it will not inject it. What do you think?

**Comment by nandorojo** [comment]
Yeah I'm thinking the API is, there is a `where[3]` field, and before stringifying, the lib checks if `where[2] instanceof Date`, and if so, makes `where[3]` into `type: date`.

Then, when creating the firestore ref, we check if we have `type: date`.

**Comment by AlexanderArvidsson** [comment]
@nandorojo I have now submitted a PR (#75) which implements a proper serializer which will allow us to easily serialize any kind of object in the future. Please take a look!

**Comment by AlexanderArvidsson** [comment]
I have also noticed that you cannot query for document references. For example:
```
  const { data: test } = useCollection<CM.Share>(
    portfolio ? 'shares' : null,
    portfolio && {
      where: ['portfolios', 'array-contains', firestore.doc(`portfolios/${portfolio.id}`)]
    }
  );
```
This is most likely for the exact same reason. I will add support for serializing document references to the PR.

**Comment by binajmen** [comment]
Hello @AlexanderArvidsson,

I'm currently stuck with date filtering, and your PR would be handy! What is the current status?

Thank you.

**Comment by AlexanderArvidsson** [comment]
@binajmen I am already using my fork in production. It's up to @nandorojo to accept the pull request. There is still a few improvements that can be done in the pull request (check the comments), but I do not have time to dive into it.

**Comment by nandorojo** [comment]
@AlexanderArvidsson now that you've been using it in prod, are you confident in the PR's ability to merge?

**Comment by samipshah100** [comment]
I am also having the same issue. Luckily found this issue quickly. 

Any fix is highly appreciated! 

Thanks.  😀

**Comment by lawtheawe** [comment]
Hello, any workarounds that I can query the timestamp by range at this moment?

**Comment by 345ml** [comment]
I ran into a problem where the process kept repeating without being able to retrieve any data.

By pushing the date out of the React Component, I think this has been resolved.
I don't understand the implementation, but just wanted to report it.

```js
const date = new Date(moment().add(-7, "days").format());

const Component = () => {
  const { data } = useCollection("users", {
    where: ["createdAt", ">=", whereDate],
  });

  return <div></div>
}
```

**Comment by nandorojo** [comment]
That’s because the date gets recreated on each render, so it would fetch every time. You probably want this:

```ts
const date = useRef(new Date(…)).current
```

and put that inside your component.

**Comment by 345ml** [comment]
@nandorojo 
I tried milliseconds with the following code, but the query returns an empty array.
`new Date(moment().add(-7, "days").format()).getTime()`

**Comment by AlexanderArvidsson** [comment]
> That’s because the date gets recreated on each render, so it would fetch every time. You probably want this:
> 
> ```ts
> const date = useRef(new Date(…)).current
> ```
> 
> and put that inside your component.
> 
> but yeah, currently this only works with a milliseconds time stamp as this issue states.

The date is created outside the component, surely it would not be re-created every render in that case unless you use lazy loading components (which I believe still wouldn't suffer much from this problem anyway?
